diff -up chrony-4.1/examples/chrony.conf.example2.defconfig chrony-4.1/examples/chrony.conf.example2
--- chrony-4.1/examples/chrony.conf.example2.defconfig	2021-05-12 13:06:15.000000000 +0200
+++ chrony-4.1/examples/chrony.conf.example2	2019-05-10 12:22:57.000000000 +0200
@@ -1,5 +1,5 @@
 # Use public servers from the pool.ntp.org project.
-# Please consider joining the pool (https://www.pool.ntp.org/join.html).
+# Please consider joining the pool (http://www.pool.ntp.org/join.html).
 pool pool.ntp.org iburst
 
 # Record the rate at which the system clock gains/losses time.
@@ -25,18 +25,9 @@ rtcsync
 # Serve time even if not synchronized to a time source.
 #local stratum 10
 
-# Require authentication (nts or key option) for all NTP sources.
-#authselectmode require
-
 # Specify file containing keys for NTP authentication.
 #keyfile /etc/chrony.keys
 
-# Save NTS keys and cookies.
-ntsdumpdir /var/lib/chrony
-
-# Insert/delete leap seconds by slewing instead of stepping.
-#leapsecmode slew
-
 # Get TAI-UTC offset and leap seconds from the system tz database.
 #leapsectz right/UTC
 
diff -up chrony-4.5/examples/chrony.keys.example.keys chrony-4.5/examples/chrony.keys.example
--- chrony-4.5/examples/chrony.keys.example.keys	2023-12-05 14:22:10.000000000 +0100
+++ chrony-4.5/examples/chrony.keys.example	2023-12-06 09:59:26.089508934 +0100
@@ -11,5 +11,3 @@
 #1 MD5 AVeryLongAndRandomPassword
 #2 MD5 HEX:12114855C7931009B4049EF3EFC48A139C3F989F
 #3 SHA1 HEX:B2159C05D6A219673A3B7E896B6DE07F6A440995
-#4 AES128 HEX:2DA837C4B6573748CA692B8C828E4891
-#5 AES256 HEX:2666B8099BFF2D5BA20876121788ED24D2BE59111B8FFB562F0F56AE6EC7246E
